### PR TITLE
CORE-4618: Keys API: List, retrieve and generate keys for cluster and member

### PR DESCRIPTION
**Pull request by yift-r3**
Currently, this won't work without a hack (https://github.com/corda/corda-runtime-os/pull/1303).

Usage example:
```
$ curl -k -u admin:admin -X POST -H "Content-Type: application/json" -d '{"alias":"a1","hsmCategory":"TLS"}' https://localhost:8888/api/v1/keys/cluster
"A641B7AEFD27"%
$ curl -k -u admin:admin -X POST -H "Content-Type: application/json" -d '{"alias":"a2","hsmCategory":"TLS"}' https://localhost:8888/api/v1/keys/cluster
"84C30ADBD045"%
$ curl -k -u admin:admin -X POST -H "Content-Type: application/json" -d '{"alias":"a3","hsmCategory":"TLS"}' https://localhost:8888/api/v1/keys/cluster
"F7AA33121828"%
$ curl -k -u admin:admin https://localhost:8888/api/v1/keys/cluster
["A641B7AEFD27","84C30ADBD045","F7AA33121828"]%
$ curl -k -u admin:admin https://localhost:8888/api/v1/keys/cluster/84C30ADBD045
"-----BEGIN PUBLIC KEY-----\nMIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEAkmHTerqWVeFDR447IQZ6\nQ05cxCOKu+oezZvgJLWMLT9sBdrT9wZNw8mfoZ2v1rmLwIH516aRXtcKPbkuPgBH\nZuXPxWPmvaKTWwNsgSsRGY9HmSVKI0C5QUFWNawn8oVQerXBqqRacKlgFM87tvdB\nngqdmOp8Rerq2CB9do2jDGuhs+xDqU9KzdiYKNtS4Ve7ZL6pBqAA3W15T93wSW45\n7HNLR9ZZRNd8UVQEV6zUjD8iXoNsQhwjbznPSZFlytPAqaOveeBZ6Pt8b/YPwchK\nevWcSwUlkEBT45NJ77+OgEy0bUdeHPbq1cDYEsMoTPaFuIDSJ4cdosXylmH6TC2H\ncobyU31quj+riBO1VDd5uNZ97pRdLoTQ2RrPooZ6mtQpX4M/AghjqqVL6LE7sMPa\n6UzZwCxeYxHI2kAkdGgJpprFiRFTBJ9s7CHcZLvMV/8nb88PmVWyyMMRMgD8gNB7\nIMH7N3HNslFSH6rihLiMZ5HHjRpEto5fgVUOBNzadj81AgMBAAE=\n-----END PUBLIC KEY-----\n"% 
```

### CODE REVIEW

**Review by charlieR3 — COMMENTED** [review]
Looks good. Two comments added.

**Review by charlieR3 — COMMENTED** [review]
A few comments left.

File: components/membership/membership-http-rpc-impl/src/main/kotlin/net/corda/membership/impl/httprpc/v1/KeysRpcOpsImpl.kt [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        ::activate,
+        ::deactivate,
+        setOf(
+            LifecycleCoordinatorName.forComponent<CryptoOpsClient>(),
+        )
+    )
+    private val coordinator = lifecycleCoordinatorFactory.createCoordinator(coordinatorName, lifecycleHandler)
+
+    override val isRunning
+        get() = coordinator.status == LifecycleStatus.UP
```
[the remark this answers is not in the document]
Comment (charlieR3) [reply components/membership/membership-http-rpc-impl/src/main/kotlin/net/corda/membership/impl/httprpc/v1/KeysRpcOpsImpl.kt:103]: Yeah, good question. I'm not a fan of `isRunning` and the way it's used across components currently as it doesn't indicate the UP/DOWN status but just whether the component is able to receive lifecycle events. 

I used to create an `isUp` property, check it on each API call, and if not up return an API error. This was not very thread safe though as I was checking `if(isUp && isRunning)`. More recently we switched to having inner implementation classes and when the component goes up we set the implementation to be an active implementation and when it goes down we set an inactive impl which returns an API error. 
e.g. MemberRegistrationRpcOps -> MemberRegistrationRpcOpsImpl -> InnerMemberRegistrationRpcOps
or MemberLookupRpcOps -> MemberLookupRpcOpsImpl -> InnerMemberLookupRpcOps
It's not really a standard across the platform but it's useful for returning proper responses or error depending on component status.

File: components/membership/membership-http-rpc-impl/src/main/kotlin/net/corda/membership/impl/httprpc/v1/KeysRpcOpsImpl.kt [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        holdingIdentityId: String,
+        alias: String,
+        hsmCategory: String,
+        scheme: String?
+    ): String {
+        return cryptoOpsClient.generateKeyPair(
+            tenantId = holdingIdentityId,
+            category = hsmCategory,
+            alias = alias,
+            scheme = scheme ?: cryptoOpsClient
```
[the remark this answers is not in the document]
Comment (alexey-kadyrov-r3) [reply components/membership/membership-http-rpc-impl/src/main/kotlin/net/corda/membership/impl/httprpc/v1/KeysRpcOpsImpl.kt]: It has to be specified explicitly

File: components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/KeysRpcOps.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+@HttpRpcResource(
+    name = "KeysRpcOps",
+    description = "Keys API",
+    path = "keys"
+)
+interface KeysRpcOps : RpcOps {
+    /**
+     * GET endpoint which returns the list of the schemes in the cluster.
+     *
+     * @param tenantId The tenant ID.
```
Comment (charlieR3) [inline components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/KeysRpcOps.kt:20]: Incomplete kdocs - missing `hsmCategory`

File: components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/KeysRpcOps.kt [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        @HttpRpcPathParameter(description = "'p2p', 'rpc-api', or holding identity identity ID.")
+        tenantId: String,
+    ): Map<String, KeyMetaData>
+
+    /**
+     * POST endpoint which Generate a key pair.
+     *
+     * @param tenantId The tenant ID.
+     * @param alias The key alias.
+     * @param hsmCategory The HSM category.
```
Comment (charlieR3) [inline components/membership/membership-http-rpc/src/main/kotlin/net/corda/membership/httprpc/v1/KeysRpcOps.kt:56]: kdocs missing `scheme`

**Review by charlieR3 — COMMENTED** [review]
Looks good. Just kdocs and still unsure about that implementation of `val isRunning`

File: components/membership/membership-http-rpc-impl/src/main/kotlin/net/corda/membership/impl/httprpc/v1/KeysRpcOpsImpl.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (yift-r3) [reply components/membership/membership-http-rpc-impl/src/main/kotlin/net/corda/membership/impl/httprpc/v1/KeysRpcOpsImpl.kt:103]: I don't like the implementation in the MemberRegistrationRpcOps, not only that it too verbose, but it's also kind of "wrong", the ops are not really running when the coordinator is running, i.e. they can't handle any request which is the only thing they meant to do while running...

**Review by charlieR3 — APPROVED** [review]

**Pull request merged**